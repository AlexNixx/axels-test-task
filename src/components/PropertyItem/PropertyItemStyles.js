import {Card} from "react-bootstrap";
import {Link} from "react-router-dom";
import styled from "styled-components";

export const StyledLink = styled(Link)`
  text-decoration: none;
`

export const StyledCard = styled(Card)`
  width: 18rem;
  height: 25rem;
  overflow: hidden;
  cursor: pointer;
  
`

export const StyledImage = styled(Card.Img)`
  width: 100%;
  height: 15rem;
  object-fit: cover;
  transition: transform 0.25s;

  &:hover {
    transform: scale(1.05);
  }
`