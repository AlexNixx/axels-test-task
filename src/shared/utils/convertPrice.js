import PropTypes from 'prop-types';

export const convertPrice = price =>
    new Intl.NumberFormat('en-US', {
        style: 'currency',
        currency: 'USD'
    }).format(price);

convertPrice.propTypes = {
    price: PropTypes.number
};
