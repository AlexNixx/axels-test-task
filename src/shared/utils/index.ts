export * from './test/renderWithProviders';
export * from './test/showLocation';
export * from './convertPrice/convertPrice';
