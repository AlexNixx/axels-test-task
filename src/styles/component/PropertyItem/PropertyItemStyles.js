import { Card } from 'react-bootstrap';
import { Link } from 'react-router-dom';
import styled from 'styled-components';

export const StyledLink = styled(Link)`
    text-decoration: none;
`;

export const StyledCard = styled(Card)`
    height: 20rem;
    overflow: hidden;
    cursor: pointer;
`;

export const StyledImage = styled(Card.Img)`
    height: 10rem;
    width: 100%;
    object-fit: cover;
    transition: transform 0.25s;

    &:hover {
        transform: scale(1.05);
    }
`;
