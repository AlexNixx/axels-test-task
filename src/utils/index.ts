export * from './test/renderWithProviders';
export * from './test/showLocation';
export * from './test/mockProperties';
export * from './convertPrice/convertPrice';
