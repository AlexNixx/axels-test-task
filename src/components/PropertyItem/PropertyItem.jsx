import {Card} from "react-bootstrap";
import {StyledCard, StyledImage, StyledLink} from "./PropertyItemStyles";
import {convertToUSAPrice} from "../../utils/price/convertToUSAPrice";

export const PropertyItem = ({id, image, title, price, address}) => {
    return (
        <StyledLink to={`/property/${id}`}>
            <StyledCard>
                <StyledImage variant="top" src={image[0]} alt={`property ${title}`}/>
                <Card.Body>
                    <Card.Title>{title}</Card.Title>
                    <Card.Subtitle>
                        {convertToUSAPrice(price)}
                    </Card.Subtitle>
                    <Card.Text>
                        {address}
                    </Card.Text>
                </Card.Body>
            </StyledCard>
        </StyledLink>
    );
}
