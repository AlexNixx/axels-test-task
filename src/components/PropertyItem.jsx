import { Card } from 'react-bootstrap';
import PropTypes from 'prop-types';

import { convertPrice } from '../shared/utils';
import {
    StyledCard,
    StyledImage,
    StyledLink
} from '../styles/component/PropertyItemStyles';

export const PropertyItem = ({ id, images, title, price, address }) => (
    <StyledLink to={`/property/${id}`}>
        <StyledCard>
            <StyledImage
                variant='top'
                src={images[0]}
                alt={`property ${title}`}
            />
            <Card.Body>
                <Card.Title>{title}</Card.Title>
                <Card.Subtitle>{convertPrice(price)}</Card.Subtitle>
                <Card.Text>{address}</Card.Text>
            </Card.Body>
        </StyledCard>
    </StyledLink>
);

PropertyItem.propTypes = {
    id: PropTypes.number,
    images: PropTypes.array,
    title: PropTypes.string,
    price: PropTypes.number,
    address: PropTypes.string
};
