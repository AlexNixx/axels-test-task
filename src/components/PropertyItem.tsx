import { Link } from 'react-router-dom';
import { Card } from 'react-bootstrap';

import { convertPrice } from '../shared/utils';
import {
    StyledCard,
    StyledImage
} from '../styles/component/PropertyItemStyles';

export interface PropertyItemProps {
    id: number;
    images: string[];
    title: string;
    price: number;
    address: string;
}

export const PropertyItem = ({
    id,
    images,
    title,
    price,
    address
}: PropertyItemProps) => (
    <Link to={`/property/${id}`} data-testid='propertyLink'>
        <StyledCard data-testid='propertyCard'>
            <StyledImage
                variant='top'
                src={images[0]}
                alt={`property ${title}`}
            />
            <Card.Body>
                <Card.Title data-testid='cardTitle'>{title}</Card.Title>
                <Card.Subtitle data-testid='cardPrice'>
                    {convertPrice(price)}
                </Card.Subtitle>
                <Card.Text data-testid='cardAddress'>{address}</Card.Text>
            </Card.Body>
        </StyledCard>
    </Link>
);
