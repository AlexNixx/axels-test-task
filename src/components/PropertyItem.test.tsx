import { act, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import { PropertyItem, PropertyItemProps } from './PropertyItem';
import { renderWithProviders, convertPrice } from '../shared/utils';

describe('Test PropertyItem Card', () => {
    let defaultProps: PropertyItemProps;
    beforeEach(() => {
        defaultProps = {
            id: 1,
            title: 'Property Title',
            price: 1000,
            address: 'Address, Kyiv',
            images: ['imageurl']
        };
    });

    test('should be successfully rendered', async () => {
        renderWithProviders(<PropertyItem {...defaultProps} />);

        expect(screen.getByTestId('propertyCard')).toBeInTheDocument();
        expect(screen.getByTestId('cardTitle')).toBeInTheDocument();
        expect(screen.getByTestId('cardPrice')).toBeInTheDocument();
        expect(screen.getByTestId('cardAddress')).toBeInTheDocument();
    });

    test('should be filled with text', async () => {
        const { container } = renderWithProviders(
            <PropertyItem {...defaultProps} />
        );

        expect(screen.getByTestId('cardTitle')).toContainHTML(
            defaultProps.title
        );

        expect(screen.getByTestId('cardPrice')).toContainHTML(
            convertPrice(defaultProps.price)
        );
        expect(screen.getByTestId('cardPrice')).not.toContainHTML(
            '' + defaultProps.price
        );

        expect(screen.getByTestId('cardAddress')).toContainHTML(
            defaultProps.address
        );

        expect(container).toMatchSnapshot();
    });

    test('should go to the detail page', async () => {
        renderWithProviders(<PropertyItem {...defaultProps} />);

        const propertyLink = screen.getByTestId('propertyLink');
        await act(async () => await userEvent.click(propertyLink));

        expect(await screen.findByTestId('propertyPage')).toBeInTheDocument();
    });
});
