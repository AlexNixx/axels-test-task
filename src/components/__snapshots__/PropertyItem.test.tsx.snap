// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PropertyItem Card should be filled with text 1`] = `
<div>
  <div
    data-testid="locationDisplay"
  >
    /
  </div>
  <a
    data-testid="propertyLink"
    href="/property/1"
  >
    <div
      class="sc-bcPKhP bZwkZu card"
      data-testid="propertyCard"
    >
      <img
        alt="property Property Title"
        class="card-img-top sc-grXZZQ dsaGvy"
        src="imageurl"
      />
      <div
        class="card-body"
      >
        <div
          class="card-title h5"
          data-testid="cardTitle"
        >
          Property Title
        </div>
        <div
          class="card-subtitle h6"
          data-testid="cardPrice"
        >
          $1,000.00
        </div>
        <p
          class="card-text"
          data-testid="cardAddress"
        >
          Address, Kyiv
        </p>
      </div>
    </div>
  </a>
</div>
`;
